[Joshua Emmanuel Teo] iP

jetrz opened the pull request:
# **Your Best Friend Pikachu!**

![Alt Text](https://c.tenor.com/9jn_TYgvSyQAAAAC/pikachu-pokemon.gif)
> #### _"Pika pika!"_

#### Pikachu helps you keep organised by recording your daily tasks and displaying them on command! :yellow_heart:

## Table of Contents
  * [Setup](#setup)
  * [Features](#features)
  * [Customisation](#customisation)

## Setup
1. Download the .jar file [here](https://github.com/jetrz/ip/releases/tag/A-Jar).
2. Open your local terminal and navigate to the folder which you saved the .jar file in, and run the following command:
```
java -jar ip.jar
```
3. That's it! Have fun with Pikachu!

## Features
- [x] To-dos, Deadline and Event task types implemented
- [x] Marking tasks as done/not done
- [x] Searching for tasks by name
- [ ] Interactive GUI (coming soon)

## Customisation
Want to customize your own Pikachu? Modify the following functions in the UI.java file to control what Pikachu says! 
```java
public static void printGreeting() {
    //Your customized greeting here!
}

public static void printCommands() {
    //Your customized command list here!
}

public static void printGoodbye() {
    //Your customized goodbye here!
}
```

Commit:
Add Gradle support

jetrz committed:
Level-1

jetrz committed:
Level-2

jetrz committed:
Level-3

jetrz committed:
Level-4

jetrz committed:
A-TextUiTesting

jetrz committed:
Level-6

jetrz committed:
Level-5

jetrz committed:
Level-6

Commit:
add save functionality

Commit:
Change file structure & abstract storage class

Added duke parent directory and packaged all files into duke and their respective sub-folders

Also abstracted storage functions out from duke.java and into Storage.java

Commit:
Add jar compatibility

Commit:
Add UI abstraction

Added UI class for abstraction of UI aesthetics

Commit:
Add date-time functionality & 'commands' command

Deadline and event now stores its respective timings as LocalDateTime instead of String. Also added new 'commands' function to print out the list of available commands and their respective syntax.

Commit:
Clean up pikachu.java

Commit:
Add javadoc comments

Added javadoc comments for all non-trivial functions.

Commit:
Add find command

Added command to allow user to search for tasks containing a given keyword.

Commit:
Merge branch 'Branch-Level-9'

Commit:
Fix coding standards

Commit:
Add JUnit testing

Added JUnit testing for Pikachu (Parser) class, testing several commands and error messages.

Commit:
Merge branch 'add-gradle-support'

zixin448 commented on src/main/java/duke/UI/UI.java:
```diff
@@ -0,0 +1,65 @@
+
+public class UI {
+
+    public static void printGreeting() {
```
This is intense haha wow!!

zixin448 commented on src/main/java/duke/tasklist/Storage.java:
```diff
@@ -0,0 +1,78 @@
+                this.pikachu.inputList.add(t);
+                if (split[1].equals("1")) t.mark(); 
+            } else if (split[0].equals("D")) {
+                LocalDateTime deadline = LocalDateTime.of(Integer.parseInt(split[3]), Integer.parseInt(split[4]), 
+                        Integer.parseInt(split[5]), Integer.parseInt(split[6]), Integer.parseInt(split[7]));
+                Deadline d = new Deadline(split[2], deadline);
+                this.pikachu.inputList.add(d);
+                if (split[1].equals("1")) d.mark(); 
+            } else if (split[0].equals("E")) {
+                LocalDateTime start = LocalDateTime.of(Integer.parseInt(split[3]), Integer.parseInt(split[4]), 
+                        Integer.parseInt(split[5]), Integer.parseInt(split[6]), Integer.parseInt(split[7]));
+                LocalDateTime end = LocalDateTime.of(Integer.parseInt(split[8]), Integer.parseInt(split[9]), 
+                        Integer.parseInt(split[10]), Integer.parseInt(split[11]), Integer.parseInt(split[12]));
```
Perhaps add some comments here to explain what the split[i] are representing?

zixin448 commented on src/main/java/duke/Pikachu.java:
```diff
@@ -0,0 +1,227 @@
+         * Syntax: event <taskname> /<yyyy-mm-dd hhmm of start> <yymmdd hhmm of end>
+         */
+        if (split[0].toLowerCase().equals("event")) {
+            try {
+            String[] split2 = str.split("/");
+            String[] split3 = split2[1].split(" ");
+            String[] split4 = split2[0].split(" ", 2);
+            String name = split4[1];
+            String dateStart = split3[0];
+            String timeStart = split3[1];
+            String dateEnd = split3[2];
+            String timeEnd = split3[3];
+            String[] yymmddStart = dateStart.split("-");
+            String[] yymmddEnd = dateEnd.split("-");
```
I like how you assigned the split[i] to their respective variables! It makes everything look neat and easy to understand!

zixin448 commented on src/main/java/duke/UI/UI.java:
```diff
@@ -0,0 +1,65 @@
+        System.out.printf("             _'....----\"\"\"\"\" \n\n");
+    }
+
+    public static void printCommands() {
+        System.out.println("The available commands are:\ncommands: See list of commands\nlist: See list of current tasks");
+        System.out.println("mark: Mark a task as completed. Syntax: mark <index of task>");
+        System.out.println("unmark: Mark a task as not completed. Syntax: unmark <index of task>");
+        System.out.println("delete: Delete a task. Syntax: delete <index of task>");
+        System.out.println("todo: Create a todo task. Syntax: todo <taskname>");
+        System.out.println("deadline: Create a deadline. Syntax: deadline <taskname> /<yyyy-mm-dd hhmm of deadline>");
+        System.out.println("event: Create an event. Syntax: event <taskname> /<yyyy-mm-dd hhmm of start> <yymmdd hhmm of end>");
+        System.out.println("find: Search for tasks containing a given keyword. Syntax: find <keyword>");
```
Maybe print all as a block of sentences instead of repeating System.out.println()?

zixin448 commented on src/main/java/duke/tasks/Event.java:
```diff
@@ -0,0 +1,43 @@
+        super(name);
+        this.start = start;
+        this.end = end;
+        this.info = "E,0," + name + "," + start.getYear() + "," + start.getMonthValue() + "," + start.getDayOfMonth()
+                + "," + start.getHour() + "," + start.getMinute() + "," + end.getYear() + "," + end.getMonthValue() + "," + 
+                end.getDayOfMonth() + "," + end.getHour() + "," + end.getMinute();
+    }
+
+    @Override
+    public void mark() {
+        this.isDone = true;
+        this.info = "E,1," + name + "," + start.getYear() + "," + start.getMonthValue() + "," + start.getDayOfMonth()
+                + "," + start.getHour() + "," + start.getMinute() + "," + end.getYear() + "," + end.getMonthValue() + "," + 
+                end.getDayOfMonth() + "," + end.getHour() + "," + end.getMinute();
+    }
```
Will it be better if these statements are abstracted into another method? 'this.info = ...' appears to have been repeated a few times!

Commit:
Fix checkstyle issues

sharpstorm commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,35 @@
+
+import duke.tasks.*;
+import duke.tasklist.*;
+import duke.UI.*;
```
Should wildcard imports be used? I noticed this issue in several other classes.

sharpstorm commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,35 @@
+        //Start accepting commands
+        while (true) {
+            currInput = sc.nextLine(); //scan in user input
+            if (currInput.toLowerCase().equals("bye")) break; //if user input == bye, exit programme
```
I believe 1-liner if clauses do not comply with the coding standard. Also, shouldn't there be curly braces for one-liner if statements?
```suggestion
            if (currInput.toLowerCase().equals("bye")) {
                break; //if user input == bye, exit programme
            }
```

sharpstorm commented on src/main/java/duke/Pikachu.java:
```diff
@@ -0,0 +1,247 @@
+         * "commands"
+         * Prints list of available commands.
+         * Syntax: commands
+         */
```
I feel that Javadocs should not be in a method body. Perhaps consider changing to comment blocks instead.
```suggestion
        /*
         "commands"
         Prints list of available commands.
         Syntax: commands
         */
```

sharpstorm commented on src/main/java/duke/Pikachu.java:
```diff
@@ -0,0 +1,247 @@
+     * @param str Input from user.
+     */
+    public void parseInput(String str) {
+        String[] split = str.split(" ");
```
Shouldn't variable name for arrays be plural? Similar issues throughout this function.
```suggestion
        String[] splitStrings = str.split(" ");
```

sharpstorm commented on src/main/java/duke/Pikachu.java:
```diff
@@ -0,0 +1,247 @@
+         */
+        if (split[0].toLowerCase().equals("deadline")) {
+            try {
+                String[] split2 = str.split("/");
+                String[] split3 = split2[1].split(" ");
+                String[] split4 = split2[0].split(" ", 2);
```
As commented above, shouldn't variable names for arrays be plural?

sharpstorm commented on src/main/java/duke/UI/UI.java:
```diff
@@ -0,0 +1,67 @@
+package duke.UI;
```
Should package name contain any capitalized alphabets?
```suggestion
package duke.ui;
```

sharpstorm commented on src/main/java/duke/UI/UI.java:
```diff
@@ -0,0 +1,67 @@
+package duke.UI;
+
+public class UI {
```
While I get the abbreviation, shouldn't class names follow the camelCase guideline?
```suggestion
public class Ui {
```

sharpstorm commented on src/main/java/duke/tasklist/Storage.java:
```diff
@@ -0,0 +1,86 @@
+            if (split[0].equals("T")) {
+                ToDo t = new ToDo(split[2]);
+                currInputList.add(t);
+                if (split[1].equals("1")) t.mark();
```
Should one-liner if-statements be used? Similar issue throughout the rest of the cases in this function.
```suggestion
                if (split[1].equals("1")) {
                    t.mark();
                }
```

sharpstorm commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,44 @@
+    private LocalDateTime deadline;
+
+    //Constructor
+    public Deadline (String name, LocalDateTime deadline) {
```
I believe there shouldn't be a space between the class name and the argument bracket.
```suggestion
    public Deadline(String name, LocalDateTime deadline) {
```

sharpstorm commented on src/main/java/duke/tasks/Event.java:
```diff
@@ -0,0 +1,55 @@
+    private LocalDateTime end;
+
+    //Constructor
+    public Event (String name, LocalDateTime start, LocalDateTime end) {
```
I believe there shouldn't be a space between the class name and the argument bracket.
```suggestion
    public Event(String name, LocalDateTime start, LocalDateTime end) {
```

sharpstorm commented on src/main/java/duke/tasks/ToDo.java:
```diff
@@ -0,0 +1,30 @@
+
+    @Override
+    public String toString() {
+        if (isDone){
```
Shouldn't there be no space between the closing bracket and the curly brace?
```suggestion
        if (isDone) {
```

jetrz committed:
Update README.md

Commit:
Add GUI for Pikachu using JavaFX.

All functions and output previously run in terminal have been translated to a custom GUI with styling such as profile pictures and icons!

Commit:
Merge branch 'branch-Level-10'

cindyangXQ commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,29 @@
+    private final Pikachu pikachu;
+    private final Storage storage;
+
+    //Constructor
```
Comments should have a space after the `//`?

cindyangXQ commented on src/main/java/duke/UI/UI.java:
```diff
@@ -0,0 +1,101 @@
+        System.out.printf("                   .'         /\n");
+        System.out.printf("                 .'          /\n");
+        System.out.printf("               ,'           /\n");
+        System.out.printf("             _'....----\"\"\"\"\" \n\n");
```
This logo is impressive.

cindyangXQ commented on src/main/java/duke/tasks/Event.java:
```diff
@@ -0,0 +1,55 @@
+    public void mark() {
+        this.isDone = true;
+        this.info = "E,1," + name + "," + start.getYear() + "," + start.getMonthValue() + "," + start.getDayOfMonth()
+                + "," + start.getHour() + "," + start.getMinute() + "," + end.getYear() + ","
```
It is a bit nitpicking but you might want to make the format consistent: move the first `","` of the second line to the first line or make the third line start with a `","`.

cindyangXQ reviewed (commented):
Sorry for the late comment. The code is very well written. I did not manage to find any errors.

Commit:
Add sorting by chronological order for list command

The list command now lists tasks by chronological order (for Deadlines and Events, while ToDos are all clumped at the bottom).

This makes the list command more intuitive as most people are concerned about tasks with a closer deadline.

* This was implemented entirely with a single PriorityQueue and a custom comparator (TaskComparator), which is found at the bottom of Pikachu.java.
* All changes are made entirely in the Pikachu.java file.
* This is so as to not over-complicate other portions of the code.

Commit:
Add assert checks

jetrz committed:
Merge pull request #1 from jetrz/branch-Assertions

Add assert checks

Commit:
Separate sort and list functions

Separated sorting the list by chronological order into its own command (i.e. sort), and list command now returns an unsorted list.

This was because auto-sorting in list command will cause the indexing to mess up w.r.t other commands (i.e. mark, delete etc.)

Minor styling changes were also made.

Commit:
Update UI.java

Commit:
Improve GUI

Included features such as background wallpaper, transparent user pictures, and text bubbles.

Commit:
Fix bug with 'find' command

Fixed the error caused when calling 'find' with no keyword specified.

jetrz committed:
Update README.md

jetrz committed:
Set theme jekyll-theme-midnight

Commit:
Update README.md for Github Pages

Commit:
Merge branch 'master' of https://github.com/jetrz/ip

Commit:
Update README.md for Github Pages

Commit:
Include sample pics for README

Commit:
Update sample image for higher quality

jetrz committed:
Update README.md

Include sample pictures

jetrz committed:
Update README.md
